core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3554,12 +3565,17 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
   )(
     f: A => ZIO[R, E, B]
   )(implicit bf: BuildFrom[Collection[A], B, Collection[B]], trace: Trace): ZIO[R, E, Collection[B]] =
-    if (as.isEmpty) ZIO.succeed(bf.fromSpecific(as)(Nil))
-    else
-      ZIO.parallelismWith {
-        case Some(n) => foreachPar(n)(as)(f)
-        case None    => foreachParUnbounded(as)(f)
-      }
+    as.size match {
```

Comment:
There are other things that depend on having the size of the collection, such as the sizing of an array used within one of the internal methods. So using `knownSize` won't work because it will return `-1` for Lists or other collections that their size is not known. At least with this PR we're ensuring that `size` is called just once, unlike before that it was being called 2 or 3 times